specs/WebResourceRequested-CustomScheme.md

```diff
@@ -0,0 +1,295 @@
+  // Set the array of origins that are allowed to use the scheme.
+  HRESULT SetAllowedOrigins([in] UINT32 allowedOriginsCount, [in] LPCWSTR* allowedOrigins);
+}
+
+// This is the ICoreWebView2EnvironmentOptions3 interface
+[uuid(ac52d13f-0d38-475a-9dca-876580d6793e), object, pointer_default(unique)]
+interface ICoreWebView2EnvironmentOptions3 : IUnknown {
+  /// Array of custom scheme registrations. The array must be deallocated with CoTaskMemFre.
+  HRESULT GetCustomSchemeRegistrations(
+      [out] UINT32* count,
+      [out] ICoreWebView2CustomSchemeRegistration*** schemeRegistrations);
+  /// Set the array of custom scheme registrations to be used.
+  HRESULT SetCustomSchemeRegistrations(
+      [in] UINT32 count,
+      [in] const ICoreWebView2CustomSchemeRegistration** schemeRegistrations);
+}
+```
+
+## .NET API
+```c#
+namespace Microsoft.Web.WebView2.Core
+{
+    // Represents the registration of a custom scheme with the CoreWebView2Environment.
+    // This allows the WebView2 app to be able to handle
+    // WebResourceRequested event for requests with the specified scheme and
+    // be able to navigate the WebView2 to the custom scheme. Once the environment
+    // is created, the registrations are valid and immutable throughout the
+    // lifetime of the associated WebView2s' browser process and any WebView2
+    // environments sharing the browser process must be created with identical
+    // custom scheme registrations, otherwise the environment creation will fail.
+    // Any further attempts to register the same scheme will fail.
+    // The URIs of registered custom schemes will be treated similar to http URIs for their origins.
+    // They will have tuple origins for URIs with host and opaque origins for
+    // URIs without host as specified in [7.5 Origin - HTML Living Standard](https://html.spec.whatwg.org/multipage/origin.html)
+    // Example:
+    // `custom-scheme-with-host://hostname/path/to/resource` has origin of `custom-scheme-with-host://hostname`
+    // `custom-scheme-without-host:path/to/resource` has origin of `custom-scheme-without-host:path/to/resource`
+    // For WebResourceRequested event, the cases of request URIs and filter URIs with custom schemes
+    // will be normalized according to generic URI syntax rules. Any non-ASCII characters will be preserved.
+    // The registered custom schemes also participate in [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) and adheres to [CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP). The app needs to set the appropriate access headers in its WebResourceRequested event handler to allow CORS requests.
+    runtimeclass CoreWebView2CustomSchemeRegistration
+    {
+        // Constructor
+        CoreWebView2CustomSchemeRegistration(string schemeName, bool isSecure, IList<String> allowedOrigins);
+
+        // The name of the custom scheme to register.
+        string SchemeName { get; set; };
```

Comment:
It's a bit strange (but not necessarily "bad") to require these things in the .ctor and then also have them be settable. 

For example, the current API invites re-use of objects; is that allowed? E.G.

```C#
var r = new Reg("foo", true, {});
regs.Add(r);
r.SchemeName = "bar";
regs.Add(r);
r.SchemeName = "baz";
regs.Add(r);
```

Reply:
Fixed see above.